src/Microsoft.OData.Core/ODataNotificationReader.cs

```diff
@@ -113,18 +121,44 @@ public override Task<string> ReadToEndAsync()
         /// <param name="disposing">True if called from Dispose; false if called from the finalizer.</param>
         protected override void Dispose(bool disposing)
         {
-            if (disposing)
+            if (!this.disposed && disposing)
             {
-                if (this.listener != null)
+                // Tell the listener that the stream is being disposed.
+                if (synchronous)
                 {
-                    // Tell the listener that the stream is being disposed.
-                    this.listener.StreamDisposed();
-                    this.listener = null;
+                    this.listener?.StreamDisposed();
                 }
+                else
+                {
+                    this.listener?.StreamDisposedAsync().Wait();
+                }
+
+                this.listener = null;
+                // NOTE: Do not dispose the text reader since this instance does not own it.
+                this.textReader = null;
             }
 
-            this.textReader.Dispose();
+            this.disposed = true;
             base.Dispose(disposing);
         }
+
+#if NETSTANDARD2_0
+        public async ValueTask DisposeAsync()
+        {
+            if (!this.disposed && this.listener != null)
+            {
+                await this.listener.StreamDisposedAsync()
+                    .ConfigureAwait(false);
+
+                this.listener = null;
+                // NOTE: Do not dispose the text reader since this instance does not own it.

```

Comment:
I'm a little nervous about this change of behavior for not disposing `textReader`. I agree with the change on principal (the `ODataNotificationReader` does not own the `textReader`) but anyone currently using the `ODataNotificationReader` is already going to be expecting it to dispose the `TextReader`. Perhaps we should add a flag in the constructor so that current consumers get the existing behavior, but future consumers are able to maintain control of their `TextReader`?

Reply:
@corranrogue9 The `ODataNotificationReader` is internal and the callers are internal. Instead of taking the longer path of introducing a flag, I'll verify and ensure that the callers dispose the `textReader` object since they own it

Reply:
I'd argue in this case (based on the usage scenarios) that the caller "transferred" ownership of the `textReader` to this class.

Reply:
@habbes True. That's the reasoning I went with in the end. I added a remark to the constructor indicating that when this class is disposed, it in turns disposes the wrapped text reader